content/automate/ManageIQ/Cloud/VM/Provisioning/StateMachines/ProvisionRequestApproval.class/__methods__/validate_request.rb

```diff
@@ -5,203 +5,225 @@
-    approval_req = true
-    reason3 = "Requested VMs #{desired_nvms} limit is #{max_vms}"
-  end
-end
-
-###################################
-#
-# max_memory:
-# Validate max_memory by first checking the below
-# value, then check the model and finally the template tag
-#
-###################################
-# Set max_memory here to override the model
-max_memory = nil
-
-# Use value from model unless specified above
-max_memory ||= $evm.object['max_memory']
-unless max_memory.nil?
-  $evm.log("info", "Auto-Approval Threshold(Model):<max_memory=#{max_memory}> detected")
-end
-
-# Reset to nil if value is zero
-max_memory = nil if max_memory == '0'
-
-# Get Tag
-prov_max_memory = template.tags(:prov_max_memory).first
-# If template is tagged then override
-unless prov_max_memory.nil?
-  $evm.log("info", "Auto-Approval Threshold(Tag):<prov_max_memory=#{prov_max_memory}> from template:<#{template.name}> detected")
-  max_memory = prov_max_memory.to_i
-end
-
-# Validate max_memory if not nil or empty
-unless max_memory.blank?
-  desired_mem = requested_memory(prov_resource, desired_nvms)
-  if desired_mem && (desired_mem.to_i > max_memory.to_i)
-    $evm.log("info", "Auto-Approval Threshold(Warning): Number of vRAM requested: \
-    <#{desired_mem.to_s(:human_size)}> exceeds:<#{max_memory}>")
-    approval_req = true
-    reason2 = "Requested Memory #{desired_mem.to_s(:human_size)} limit is #{max_memory}"
+module ManageIQ
+  module Automate
+    module Cloud
+      module VM
+        module Provisioning
+          module StateMachines
+            module ProvisionRequestApproval
+              class ValidateRequest
+                def initialize(handle = $evm)
+                  @handle = handle
+                end
+
+                def main
+                  initialize_variables
+                  validate_max_cpus
+                  validate_max_vms
+                  validate_max_memory
+                  validate_max_retirement_days
+
+                  # Update Message to Requester
+                  if @approval_req == true
+                    msg =  "Request was not auto-approved for the following reasons: "
+                    msg += "(#{@reason1}) " if @reason1
+                    msg += "(#{@reason2}) " if @reason2
+                    msg += "(#{@reason3}) " if @reason3
+                    msg += "(#{@reason4}) " if @reason4
+                    @prov_resource.set_message(msg)
+                    @handle.log("info", "Auto-Approval #{msg}")
+
+                    @handle.root['ae_result'] = 'error'
+                    @handle.object['reason'] = msg
+                  end
+                end
+
+                private
+
+                def initialize_variables
+                  @prov = @handle.root['miq_request']
+                  @prov_resource = @prov.resource
+                  raise "Provisioning Request not found" if @prov.nil? || @prov_resource.nil?
```

Comment:
@pkomanek 
Since this is related to requests and tasks would it make sense to have another method in
https://github.com/ManageIQ/manageiq-content/blob/a31fac6d8787c2c5ca27ea24a261fb2c4a9e3d86/content/automate/ManageIQ/System/CommonMethods/Utils.class/__methods__/log_object.rb#L56

that can be called log_and_task(...,msg,task) log_and_task_and_raise(...., msg, task) log_and_request(...., msg, request) or log_and_request_raise(..,msg, request)

Because we dont want to loose these error conditions in the log, we can have it be stored in the request as well as in the task